pkg/dns/treecache.go

```diff
@@ -0,0 +1,312 @@
+limitations under the License.
+*/
+
+package dns
+
+import (
+	"bytes"
+	"crypto/md5"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path"
+	"reflect"
+	"strings"
+	"sync"
+)
+
+const (
+	dataFile = "data.dat"
+	crcFile  = "data.crc"
+)
+
+type object interface{}
+
+type TreeCache struct {
+	ChildNodes map[string]*TreeCache
+	Entries    map[string]interface{}
+	m          *sync.RWMutex
+}
+
+func NewTreeCache() *TreeCache {
+	return &TreeCache{
+		ChildNodes: make(map[string]*TreeCache),
+		Entries:    make(map[string]interface{}),
+		m:          &sync.RWMutex{},
+	}
+}
+
+func Deserialize(dir string) (*TreeCache, error) {
+	b, err := ioutil.ReadFile(path.Join(dir, dataFile))
+	if err != nil {
+		return nil, err
+	}
+	var hash []byte
+	hash, err = ioutil.ReadFile(path.Join(dir, crcFile))
+	if err != nil {
+		return nil, err
+	}
+	if !reflect.DeepEqual(hash, getMD5(b)) {
+		return nil, fmt.Errorf("Checksum failed")
+	}
+
+	var cache TreeCache
+	err = json.Unmarshal(b, &cache)
+	if err != nil {
+		return nil, err
+	}
+	cache.m = &sync.RWMutex{}
+	return &cache, nil
+}
+
+func (cache *TreeCache) Serialize(dir string) (string, error) {
+	cache.m.RLock()
+	defer cache.m.RUnlock()
+	b, err := json.Marshal(cache)
+	if err != nil {
+		return "", err
+	}
+
+	if len(dir) == 0 {
+		var prettyJSON bytes.Buffer
+		err = json.Indent(&prettyJSON, b, "", "\t")
+
+		if err != nil {
+			return "", err
+		}
+		return string(prettyJSON.Bytes()), nil
+	}
+	if err := ensureDir(dir, os.FileMode(0755)); err != nil {
+		return "", err
+	}
+	if err := ioutil.WriteFile(path.Join(dir, dataFile), b, 0644); err != nil {
+		return "", err
+	}
+	if err := ioutil.WriteFile(path.Join(dir, crcFile), getMD5(b), 0644); err != nil {
+		return "", err
+	}
+	return string(b), nil
+}
+
+func (cache *TreeCache) SetEntry(key string, val interface{}, path ...string) {
+	cache.m.Lock()
+	defer cache.m.Unlock()
+	node := cache.ensureChildNode(path...)
+	node.Entries[key] = val
+}
+
+func (cache *TreeCache) ReplaceEntries(entries map[string]interface{}, path ...string) {
+	cache.m.Lock()
+	defer cache.m.Unlock()
+	node := cache.ensureChildNode(path...)
+	node.Entries = make(map[string]interface{})
+	for key, val := range entries {
+		node.Entries[key] = val
+	}
+}
+
+func (cache *TreeCache) GetSubCache(path ...string) *TreeCache {
+	childCache := cache
+	for _, subpath := range path {
+		childCache = childCache.ChildNodes[subpath]
+		if childCache == nil {
+			return childCache
+		}
+	}
+	return childCache
+}
+
+func (cache *TreeCache) SetSubCache(key string, subCache *TreeCache, path ...string) {
+	cache.m.Lock()
+	defer cache.m.Unlock()
+	node := cache.ensureChildNode(path...)
+	node.ChildNodes[key] = subCache
+}
+
+func (cache *TreeCache) GetEntry(key string, path ...string) (interface{}, bool) {
+	cache.m.RLock()
+	defer cache.m.RUnlock()
+	childNode := cache.GetSubCache(path...)
+	val, ok := childNode.Entries[key]
+	return val, ok
```

Comment:
Wherever an interface is involved in the return values, it totally scares me to use `nil` to mean anything. Thanks to Go's your-nil-pointer-to-an-underlying-concrete-type-is-not-the-same-as-nil-interface subtlety. It's probably Ok in this case because Entries' value is an interface as well, but I would stay away from that pattern wherever I can.